File: backend/bookCar/src/main/java/com/freelance/bookCar/services/invoiceService/InvoiceService.java
package com.freelance.bookCar.services.invoiceService;

import com.freelance.bookCar.dto.request.invoice.CreateInvoiceRequest;
import com.freelance.bookCar.dto.request.invoice.UpdateInvoiceRequest;
import com.freelance.bookCar.dto.response.invoiceDTO.CreateInvoiceResponse;
import com.freelance.bookCar.dto.response.invoiceDTO.GetInvoiceResponse;
import com.freelance.bookCar.dto.response.invoiceDTO.UpdateInvoiceResponse;
import com.freelance.bookCar.models.booking.Booking;
import com.freelance.bookCar.models.booking.BookingDetail;
import com.freelance.bookCar.models.invoice.Invoice;

import java.util.List;

public interface InvoiceService {
    CreateInvoiceResponse create(CreateInvoiceRequest createInvoiceRequest);
    UpdateInvoiceResponse update(UpdateInvoiceRequest updateInvoiceRequest);
    GetInvoiceResponse findById(Integer id);
    Invoice convertBookingToInvoice(Booking booking, List<BookingDetail> bookingDetails);
}

File: frontend/Angular/src/app/services/invoice/invoice.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { CreateInvoiceRequest } from '../../models/request/invoice/create-invoice-request';
import { map, Observable } from 'rxjs';
import { CreateInvoiceResponse } from '../../models/response/invoice/create-invoice-response';
import { Apiresponse } from '../../models/response/apiresponse';
import { GetInvoiceResponse } from '../../models/response/invoice/get-invoice-response';
import { UpdateInvoiceRequest } from '../../models/request/invoice/update-invoice-request';
import { UpdateInvoiceResponse } from '../../models/response/invoice/update-invoice-response';

@Injectable({
  providedIn: 'root'
})
export class InvoiceService {

  constructor(private httpClient: HttpClient) { }

  private baseUrl = 'http://localhost:8080/invoice';

  createInvoice(createInvoiceRequest: CreateInvoiceRequest): Observable<CreateInvoiceResponse> {
    return this.httpClient.post<Apiresponse<CreateInvoiceResponse>>(`${this.baseUrl}`, createInvoiceRequest).pipe(
      map((response: Apiresponse<CreateInvoiceResponse>) => {
        if (response.success) {
          return response.data;
        } else {
          throw new Error(response.message);
        }
      })
    );
  }

  updateInvoice(updateInvoiceRequest: UpdateInvoiceRequest): Observable<UpdateInvoiceResponse> {
    return this.httpClient.put<Apiresponse<UpdateInvoiceResponse>>(`${this.baseUrl}`, updateInvoiceRequest).pipe(
      map((response: Apiresponse<UpdateInvoiceResponse>) => {
        if (response.success) {
          return response.data;
        } else {
          throw new Error(response.message);
        }
      })
    );
  }

  getInvoice(id: number): Observable<GetInvoiceResponse> {
    return this.httpClient.get<Apiresponse<GetInvoiceResponse>>(`${this.baseUrl}?id=${id}`).pipe(
      map((response: Apiresponse<GetInvoiceResponse>) => {
        if (response.success) {
          return response.data; // Trả về dữ liệu nếu thành công
        } else {
          throw new Error(response.message); // Ném lỗi nếu không thành công
        }
      })
    );
  }
}
